[Fiber] disableNewFiberFeatures bugfix: host component children arrays

acdlite opened the pull request:
When `disableNewFiberFeatures` was enabled, host component children arrays were treated as empty.

sebmarkbage commented:
Doesn't this mean that fragments are not disabled?

acdlite commented:
No there's an invariant for composites. Here's the test: https://github.com/acdlite/react/blob/ff6347a270605653c1859e58ad9b30cc557ee1a4/src/renderers/dom/fiber/__tests__/ReactDOMFiber-test.js#L1027

sebmarkbage commented:
Also, what about the `<div>Text</div>` case? This seems to disable single text children, no?

What exactly does this branch do?

acdlite commented:
Haha you're right. Ok I'm definitely going to switch this flag to do be the default so we run all the tests against it.

acdlite commented:
Updated. I moved the top-level invariant to the DOM renderer because the requirements vary too much. E.g. we need to allow `null` inside the reconciler because that's how we unmount. Also, React Art supports arrays at the root.

sebmarkbage commented on src/renderers/shared/fiber/ReactChildFiber.js:
```diff
@@ -1127,42 +1126,60 @@ function ChildReconciler(shouldClone, shouldTrackSideEffects) {
+            const instance = returnFiber.stateNode;
+            if (instance.render._isMockFunction) {
+              // We allow auto-mocks to proceed as if they're
+              // returning null.
```
What you're really saying here is that they're allowed to return anything (including strings, arrays etc.)

acdlite replied:
Good point. Looks like Stack only accepts `undefined`. I'll add a test. https://github.com/facebook/react/blob/3bc5595dfd6db741b19f17e7aef98b2c1e4be847/src/renderers/shared/stack/reconciler/ReactCompositeComponent.js#L1203

acdlite replied:
Updated

acdlite commented:
@sebmarkbage @spicyj Ping

sebmarkbage commented on src/renderers/shared/fiber/ReactChildFiber.js:
```diff
@@ -1127,42 +1126,60 @@ function ChildReconciler(shouldClone, shouldTrackSideEffects) {
-                // returning null.
-                break;
-              }
+      switch (returnFiber.tag) {
```
Seems unnecessary to fork this whole function now. Can't you just add this conditional down below behind a smaller `if (ReactFeatureFlags.disableNewFiberFeatures) {` instead of forking the whole function?

acdlite replied:
Forking the entire function was the only way I could get it down to a single check. But you're right that there's a lot of overlap and it's weird to fork the entire thing. I split it out into three checks instead.
